privacyidea_radius.pm

```diff
@@ -462,17 +489,103 @@ sub authenticate {
             &radiusd::radlog( Info, "privacyIDEA Result status is true!" );
             $RAD_REPLY{'Reply-Message'} = $decoded->{detail}{message};
             if ( $decoded->{detail}{transaction_id} ) {
+                $RAD_REPLY{'State'} = $decoded->{detail}{transaction_id};
                 ## we are in challenge response mode:
-                ## 1. split the response in fail, state and challenge
+                ## OTP
+                ## 1. split the response into fail, state and challenge
                 ## 2. show the client the challenge and the state
                 ## 3. get the response and
-                ## 4. submit the response and the state to linotp and
+                ## 4. submit the response and the state to privacyIDEA and
                 ## 5. reply ok or reject
-                $RAD_REPLY{'State'} = $decoded->{detail}{transaction_id};
-                $RAD_CHECK{'Response-Packet-Type'} = "Access-Challenge";
-                # Add the response hash to the Radius Reply
-				%RAD_REPLY = ( %RAD_REPLY, mapResponse($decoded));
-                $g_return  = RLM_MODULE_HANDLED;
+                ##
+                ## PUSH
+                ## 1. split the response into fail, state and challenge
+                ## 2. poll endpoint for challenge
+                ## 3. reply ok or reject if expired
+                my $token_type = $decoded->{detail}{type};
+                if ($token_type eq "push") {
+                    # perform periodic polling for response
+                    my $token_serial = $decoded->{detail}{serial};
+                    $RAD_REPLY{'Reply-Message'} = "privacyIDEA token challenge poll failed";
+                    $g_return = RLM_MODULE_FAIL;
+
+                    # retrieve an authorization token from privacyIDEA
+                    %params = ();
+                    $params{"username"} = $PIUSER;
+                    $params{"password"} = $PIPASS;
+                    $starttime = [gettimeofday];
+                    $response = $ua->post( $AUTHURL, \%params );
+                    $content  = $response->decoded_content();
+                    $elapsedtime = tv_interval($starttime);
+                    &radiusd::radlog( Info, "elapsed time for privacyidea call: $elapsedtime" );
+                    if ( $debug == true ) {
+                        &radiusd::radlog( Debug, "Auth Content $content" );
+                    }
+                    $decoded = $coder->decode($content);
+                    if ($response->is_success && $decoded->{result}{value}) {
+                        my $auth_token = $decoded->{result}{value}{token};
+                        $ua->default_header('Authorization' => $auth_token);
+
+                        # periodic poll for challenges
+                        my $continue_poll = true;
+                        my $challenge_found = false;
+                        do {
+                            sleep(5);
```

Comment:
You are blocking the RADIUS request in the freeradius server.
So the VPN is still waiting for a RADIUS response from the RADIUS server and the UDP request will soon time out.

How would this be better than having privacyIDEA itself using ``push_wait``?